dbms/src/Functions/LeastGreatest.h

```diff
@@ -120,5 +120,325 @@ class FunctionVectorizedLeastGreatest : public IFunction
+    }
+
+    static void constantConstant(
+        const std::string & a,
+        const std::string & b,
+        const TiDB::TiDBCollatorPtr & collator,
+        std::string & c)
+    {
+        LeastGreatestStringImpl<least, use_collator>::process(collator, a, b, c);
+    }
+};
+
+
+template <bool least>
+class FunctionLeastGreatestString : public IFunction
+{
+public:
+    static constexpr auto name = least ? "tidbLeastString" : "tidbGreatestString";
+    explicit FunctionLeastGreatestString() = default;
+
+    static FunctionPtr create(const Context & context [[maybe_unused]])
+    {
+        return std::make_shared<FunctionLeastGreatestString>();
+    }
+
+    String getName() const override { return name; }
+    bool isVariadic() const override { return true; }
+
+    void setCollator(const TiDB::TiDBCollatorPtr & collator_) override { collator = collator_; }
+
+    size_t getNumberOfArguments() const override { return 0; }
+    bool useDefaultImplementationForNulls() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() < 2)
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be at least 2.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        for (const auto & argument : arguments)
+        {
+            if (!argument->isString())
+            {
+                throw Exception(
+                    fmt::format("argument type not string"),
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            }
+        }
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        size_t num_arguments = arguments.size();
+        if (num_arguments < 2)
+        {
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be at least 2.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+        }
+        if (collator)
+        {
+            executeInternal<true>(block, arguments, result);
+        }
+        else
+        {
+            executeInternal<false>(block, arguments, result);
+        }
+    }
+    template <bool use_collator>
+    void executeInternal(Block & block, const ColumnNumbers & arguments, size_t result) const
+    {
+        size_t num_arguments = arguments.size();
+
+        using impl = StringOperationImpl<least, use_collator>;
+        const auto * c0 = block.getByPosition(arguments[0]).column.get();
+        auto c_res = ColumnString::create();
+        for (size_t i = 1; i < num_arguments; ++i)
+        {
+            if (i > 1)
+            {
+                c_res = ColumnString::create();
```

Comment:
Try avoid create new column every time?

Reply:
@windtalker reminds me that we can refer to the `concat` implementation. https://github.com/pingcap/tiflash/blob/a2979078c9548de22697863e4beb2adc42afb89d/dbms/src/Functions/FunctionsString.cpp#L1205-L1211
It uses vectorized implementation for 2 arguments and polymorphic implementation for more arguments.